File: dynamic programming/1d-dp/frogJumpKSteps.java
import java.util.Arrays;
// Frog Jump with K Steps
// This is a solution for the Frog Jump problem with K steps using dynamic programming.
// The problem is to find the minimum cost for a frog to jump from the first stone to the last stone,
// where the cost of jumping from stone i to stone j is given by the absolute difference in their heights.
// The frog can jump up to K stones at a time.
// The code below implements a recursive approach to solve the problem.
// Time Complexity: O(n * k) where n is the number of stones and k is
// Space Complexity: O(n) for recursion stack space
// Recursive Approach
class Solution1 {
    private int func(int[] heights, int ind, int k) {
        if(ind == 0) return 0;

        int minSteps = Integer.MAX_VALUE;

        for(int j = 1; j <= k; j++) {
            if(ind - j >= 0) {
                int jump = func(heights, ind - j, k) + Math.abs(heights[ind] - heights[ind - j]);
                minSteps = Math.min(minSteps, jump);
            }
        }

        return minSteps;
    }
    public int frogJump(int[] heights, int k) {
        int ind = heights.length - 1;

        return func(heights, ind, k);
    }
}

// Memoization Approach
// Time Complexity: O(n * k)
// Space Complexity: O(n) for recursion stack space + O(n * k) for memoization
class Solution2 {
    private int func(int[] heights, int[] dp, int ind, int k) {
        if(ind == 0) return 0;

        if(dp[ind] != -1) return dp[ind];

        int minSteps = Integer.MAX_VALUE;

        for(int j = 1; j <= k; j++) {
            if(ind - j >= 0) {
                int jump = func(heights, dp, ind - j, k) + Math.abs(heights[ind] - heights[ind - j]);
                minSteps = Math.min(minSteps, jump);
            }
        }

        return dp[ind] = minSteps;
    }
    
    public int frogJump(int[] heights, int k) {
        int ind = heights.length - 1;
        int[] dp = new int[ind + 1];
        Arrays.fill(dp, -1);

        return func(heights, dp, ind, k);
    }
}

// Tabulation Approach
// Time Complexity: O(n * k)
// Space Complexity: O(n * k) for dp array
class Solution3 {
    public int frogJump(int[] heights, int k) {
        int ind = heights.length;
        int[] dp = new int[ind];
        Arrays.fill(dp, -1);
        dp[0] = 0;

        for(int i = 1; i < ind; i++) {
            int minSteps = Integer.MAX_VALUE;

            for(int j = 1; j <= k; j++) {
                if(i - j >= 0){
                    int jump = dp[i - j] + Math.abs(heights[i] - heights[i - j]);
                    minSteps = Math.min(minSteps, jump);
                }
            }
            dp[i] = minSteps;
        }

        return dp[ind - 1];
    }
}

// Space Optimization Approach
// Time Complexity: O(n * k)
// Space Complexity: O(1) for variables
class Solution4 {
    public int frogJump(int[] heights, int k) {
        int ind = heights.length;
        int[] dp = new int[k];

        for(int i = 1; i < ind; i++) {
            int minSteps = Integer.MAX_VALUE;

            for(int j = 1; j <= k; j++) {
                if(i - j >= 0){
                    int preInd = (i - j) % k;
                    int jump = dp[preInd] + Math.abs(heights[i] - heights[i - j]);
                    minSteps = Math.min(minSteps, jump);
                }
            }
            dp[i % k] = minSteps;
        }

        return dp[(ind - 1) % k];
    }
}

File: dynamic programming/1d-dp/frogJumpKSteps.go
package main
// frogJump calculates the minimum cost for a frog to jump from the first stone to the last stone.
// The cost of jumping from stone i to stone j is given by the absolute difference in their heights.
// The frog can jump up to k stones at a time.
// This function uses a recursive approach to achieve the solution.
// leetcode: https://leetcode.com/problems/frog-jump/
// Time Complexity: O(n * k) where n is the number of stones and k is the maximum jump length
// Space Complexity: O(n) for recursion stack space
// Recursive Approach
func funcRec(heights []int, ind int, k int) int {
    if ind == 0 {
        return 0
    }

    minStep := int(^uint(0) >> 1)

    for j := 1; j <= k; j++ {
        if ind - j >= 0 {
            diff := heights[ind] - heights[ind-j]
            if diff < 0 {
                diff = -diff
            }

            jump := funcRec(heights, ind - j, k) + diff

            if jump < minStep {
                minStep = jump
            }
        }
    }

    return minStep
}

func frogJump(heights []int, k int) int {
   ind := len(heights) - 1;

   return funcRec(heights, ind, k)
}

// memoization Approach
func funcRec2(heights []int, dp []int, ind int, k int) int {
    if ind == 0 {
        return 0
    }

    if dp[ind] != -1 {
        return dp[ind]
    }

    minStep := int(1e9)

    for j := 1; j <= k; j++ {
        if ind - j >= 0 {
            diff := heights[ind] - heights[ind-j]
            if diff < 0 {
                diff = -diff
            }

            jump := funcRec2(heights, dp, ind - j, k) + diff

            if jump < minStep {
                minStep = jump
            }
        }
    }

    dp[ind] = minStep
    return minStep
}

func frogJump2(heights []int, k int) int {
   ind := len(heights) - 1;
   dp := make([]int, ind + 1)
   for i := range dp {
    dp[i] = -1
   }

   return funcRec2(heights, dp, ind, k)
}

// tabulation Approach
func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func frogJump3(heights []int, k int) int {
	ind := len(heights)
	dp := make([]int, ind)
	for i := range dp {
		dp[i] = -1
	}
	dp[0] = 0

	for i := 1; i < ind; i++ {
		minSteps := int(1e9)

		for j := 1; j <= k; j++ {
			if i-j >= 0 {
				jump := dp[i-j] + abs(heights[i]-heights[i-j])
                if jump < minSteps {
				minSteps = jump
			}
			}
		}
        dp[i] = minSteps
	}
	return dp[ind-1]
}

// space optimized Approach
// Time Complexity: O(n * k)
// Space Complexity: O(1)
func abs2(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func frogJump4(heights []int, k int) int {
	ind := len(heights)
	dp := make([]int, k)

	for i := 1; i < ind; i++ {
		minSteps := int(1e9)

		for j := 1; j <= k; j++ {
			if i-j >= 0 {
                prevInd := (i - j) % k;
				jump := dp[prevInd] + abs2(heights[i]-heights[i-j])
                if jump < minSteps {
				minSteps = jump
			}
			}
		}
        dp[i % k] = minSteps
	}
	return dp[(ind-1) % k]
}



